### DISCUSSION TITLE
"Old" way of generating invoice IDs?

### DISCUSSION THREAD

**Discussion by VOOM108** [kimai/kimai#2636] [Questions & Answers]
With version 1.1 I used the regular invoice-IDs consisting of customername and issue-date. That worked well, because I could relate any ID to the customer at a glance and the ID would be unique for sure. 

For the newest version I can't seem to find a way to use anything but dates and incremental numbers etc. Is there a way to achieve what I have been doing so far?

**Comment by kevinpapst** [comment]
Read https://www.kimai.org/documentation/invoices.html

> For the newest version I can't seem to find a way to use anything but dates and incremental numbers etc. Is there a way to achieve what I have been doing so far?

What did you try? What results appeared? What is your desired result?

**Reply by VOOM108** [reply]
As I wrote: I still want the customername included in the ID (I used abbrev. with 4 letters max) with a result like "ABCD-210630" with ABCD being the respective customername. 

The page you linked and which I already read states: "You can mix arbitrary characters and the allowed replacer from the list below." The "list below" only lists dates and incremental numbers...

**Reply by kevinpapst** [reply]
AFAIR Kimai never created invoice numbers like ABCD-210630 but only 210630.
How did you configure that? I can't remember removing a feature. But 1.1 is 2 years old ...

Kimai does not yet support a replacer pattern for the customer name (as you found out). 
You could create your own invoice number generator. 
Or you edit the invoice number pattern and add the customer name manually before creating an invoice.
You can add a feature request for the new pattern

**Reply by VOOM108** [reply]
Thanks, that actually helped. Now I remembered, what I did.... ;) 

`#${customer.name}-${invoice.number}`

in the template.... ;)

**Reply by kevinpapst** [reply]
Aha, that explains it.

Here you go, available in the next version will be `{cname}` and `{cnumber}`: https://github.com/kevinpapst/kimai2/pull/2640/files

You could even add these few lines to your installation and use it right away (no cache clear needed 😁).

**Reply by VOOM108** [reply]
Thank you, I implemented this solution as it is more flexible.... :) 

I would like to point out, (being between 1.1 and 1.14 now) that the development is always really going into the right direction. Also with my mind-block-frustrations concerning permissions gone and with the fresh installation being updatable now even for me ;) I can finally actually profit from the constant development instead of to never change a running system.... :) So: thank you!

**Reply by kevinpapst** [reply]
Thanks for your friendly feedback! 
Hope you enjoy all the new features (which must be a lot as you missed 2 years of development)

**Reply by VOOM108** [reply]
My favorite: uploading invoice templates, obviously ;) Then I now generally understand the structure and hierarchy of settings better. The whole handling of invoices within the GUI is nice, too, I will now actually transfer the handling of due dates and paid/unpaid bills to kimai. The downloaded invoices now get a useful name with the customer and invoice-number instead of the name of the template. Very helpful, as sorting out template.docx and template (1).docx etc. was kind of a nuisance. All kinds of useful views are new within the administration of customers, activities, projects, invoices, entries etc. And there seems to be a batch-editing of entries which I have not tried yet.... 

One thing I wondered though: a new invoice gets the status "new" and the status has to be changed manually per item to "unpaid". At least in my scenario I would expect any new invoice to best be status:unpaid. Or is that status meant only for overdue payments?

**Reply by kevinpapst** [reply]
And if you update more often, it is much simpler as well.

The "new" vs "pending" state is meant to differentiate between:
- you clicked "create invoice"
- you have actually sent the invoice to your customer

As this is a two step process for now (because Kimai does not yet mail the invoice) there is this differentiation.
And it actually prevents that you forget to sent the invoice to your customer - at least it did for me :)

**Reply by VOOM108** [reply]
Ok, makes sense. In my case: I generate all invoices the same day each month and mail them instantly. So there is no real gap. That's because it's about ongoing services and not about projects billable after completion. But I don't mind the extra clicks, I was just interested what's the reason for that workflow. 

"does **not yet** mail the invoice" sounds good! :)

**Reply by kevinpapst** [reply]
That was on purpose ;-) but once step after another. some other features need to be build into the invoicing first, before than can be added.
